runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -6327,83 +6327,150 @@ static bool graDepsConflictWithInstanceOfDeps(TR::Node * depNode, TR::Node * nod
     * Last bit of cached objectClass will set to 1 indicating false cast
+    *
+    * We can request the snippet size of power 2. Following Table summarizes bytes needed for corresponding number of cache slots.
+    * 
+    * Following is the table for the number of bytes in snippet needed by each of the Cases mentioned above
+    *
+    * Number Of Slots | Case 1A | Case 1B | Case 2A | Case 2B |
+    *       1         |    8    |   16    |    4    |    8    |
+    *       2         |    16   |   64    |    16   |    32   |
+    *       3         |    32   |   64    |    16   |    32   |
+    *       4         |    64   |   128   |    32   |    64   |
+    *       5         |    64   |   128   |    32   |    64   |
+    *       6         |    64   |   128   |    32   |    64   |
+    *
     */
-   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * addressSize) + (addressSize * (maxOnsiteCacheSlots != 1));
+   
+   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * fieldSize) + (fieldSize * (maxOnsiteCacheSlots != 1) * (cacheCastClass ? 2 : 1));
+   TR::Register *dynamicCacheReg = NULL;
+   
    if (generateDynamicCache)
       {
       TR::S390WritableDataSnippet *dynamicCacheSnippet = NULL;
-      /* We can only request the snippet size of power 2, following table summarizes bytes needed for corresponding number of cache slots
-       * Case 1 : Cast class is runtime variable
-       * Case 2 : Cast class is interface / unresolved
-       * Number Of Slots |  Bytes needed for Case 1 | Bytes needed for Case 2
-       *        1        |              16          |           8
-       *        2        |              64          |           32
-       *        3        |              64          |           32
-       *        4        |              128         |           64
-       *        5        |              128         |           64
-       *        6        |              128         |           64
-       */
       int32_t requestedBytes = 1 << (int) (log2(snippetSizeInBytes-1)+1);
-      traceMsg(comp, "Requested Bytes = %d\n",requestedBytes);
-      // NOTE: For single slot cache, we initialize snippet with addressSize (4/8) assuming which can not be objectClass
-      // In all cases, we use first addressSize bytes to store offset of the circular buffer and rest of buffer will be initialized with 0xf.
+      traceMsg(comp, "Number Of Dynamic Cache Slots = %d, Caching CastClass: %s\n"
+                     "Bytes needed for Snippet = %d, requested Bytes = %d\n",maxOnsiteCacheSlots, cacheCastClass ? "true" : "false", snippetSizeInBytes, requestedBytes);
+
       TR_ASSERT_FATAL(maxOnsiteCacheSlots <= 7, "Maximum 7 slots per site allowed because we use a fixed stack allocated buffer to construct the snippet\n");
-      UDATA initialSnippet[16] = { static_cast<UDATA>(addressSize) };
+      j9objectclass_t initialSnippet[16] = { 0 };
```

Comment:
Yes, `j9objectclass_t` is not a useful type any more. It's only around to keep DDR happy at this point.

Reply:
@gacholio  I needed a type which I can use to denote the correct header is set in the snippet. Snippet is holding the J9Class pointers.

Reply:
There is no type - class pointers are either 32 or 64 bits, depending on a runtime check. Is there a reason not to use an array of `J9Class*`? If you're getting the classes using the `J9OBJECT_CLAZZ` macro, that's what it's going to return anyway.

Reply:
It is guaranteed that pointer to J9Classwill fit in 32 Bits, but on 64-Bit arch, isn't the size of pointer 64 Bit. So if I use an array of `J9Class*` , each element will take 64-Bits.

Reply:
This is a fixed-size array, so I see no reason to try and store the class pointers in their compressed format. There's no way to get your hands on a `j9objectclass_t` - they only exist in the obsolete header structs, which you can not use.

The only valid way to get the class of an object is to use the macro, which returns the uncompressed direct pointer to the `J9Class`.

Reply:
> This is a fixed-size array, so I see no reason to try and store the class pointers in their compressed format.

Reason I chose to store in compressed format is to use an instruction to load 8 byte from cache which is definitely quicker than loading quadword from memory in single instruction.

Reply:
I have changed the array type to `U_32` instead of `j9objectclass_t` in https://github.com/eclipse/openj9/pull/9517/commits/eacc941c50cd958900914c934e54ebeeac8a0e6a
I don't expect any errors or failures but would like to throw this under test once again. Will update the PR once test passes.

Reply:
To support mixed mode, you're going to need to support both at runtime, so the array will need to be UDATA sized anyway. For the compressed case, you could store compressed class pointers there (leaving the last half of the array unused).

https://github.com/eclipse/openj9/blob/2ac59e97f9e40755f54c3cbdee417241d5f93afb/runtime/compiler/z/runtime/s390_macros.inc#L560-L602

Reply:
The U32-sized array will certainly fail on 64-bit non-compressed.

Reply:
> The U32-sized array will certainly fail on 64-bit non-compressed.

This array is just used for initialization of data in snippet. Snippet is used as circular buffer so header is used for storing the offset of update slot. On 64-Bit Non Compressed pointers,
Instead of storing offset in first 32-Bit, it will store it in second 32-Bit. On runtime we load the 64-Bit value from the header so I do not expect it to fail.

Reply:
> To support mixed mode, you're going to need to support both at runtime, so the array will need to be UDATA sized anyway. For the compressed case, you could store compressed class pointers there (leaving the last half of the array unused).

Ok, I will have to make changes to support runtime compressed refs. I can open up the issue to keep track of the changes as we are close to 0.21 release and I wanted to get this fixes in to get enough testing. 
@fjeremic  What do you suggest? I believe to support runtime compressed refs, fix will need to  LPQ/STPQ for 64 Bit platform and LG/STG for 31 Bit (In case of dynamic cast class) but that I will need to test the changes.

Reply:
> @fjeremic What do you suggest? I believe to support runtime compressed refs, fix will need to LPQ/STPQ for 64 Bit platform and LG/STG for 31 Bit (In case of dynamic cast class) but that I will need to test the changes.

I don't think you understood that correctly. The choice of whether the JVM will run in compressedrefs mode vs. full 64-bit pointers is changing to be dynamic at JVM startup, rather than having two separate JVMs to be able to do this. From the compiler perspective nothing changes on our end because the choice of compressedrefs vs. not is made at JVM initialization and it is never changed after that. The code is fine as is.

The only problem is use of that `j9objectclass_t` type which today has a different size on compressedrefs JVM vs. full 64-bit pointer JVM, but that will change as soon as we live in a world where those JVMs are one and the same. You need to ensure correct size of the snippet in this world, and it cannot rely on the `sizeof(j9objectclass_t)` at compile time. IMO just use `uintptr_t` or `UDATA`. You will end up allocating double the size when running in compressedrefs but who cares, it's just a stack object we temporarily use and feed into `cg->CreateConstant`.

Reply:
> The only problem is use of that j9objectclass_t type which today has a different size on compressedrefs JVM vs. full 64-bit pointer JVM, but that will change as soon as we live in a world where those JVMs are one and the same. You need to ensure correct size of the snippet in this world, and it cannot rely on the sizeof(j9objectclass_t) at compile time. IMO just use uintptr_t or UDATA. You will end up allocating double the size when running in compressedrefs but who cares, it's just a stack object we temporarily use and feed into cg->CreateConstant.

Ohh, Got that.  So current solution uses `useCompressedPointers` query to know if it is compressed refs or not. In last commit I have changed the `j9objectclass_t` to U_32 and update the second element with the update offset if header needs to be 8 bit, else updates first element. 
Nothing changed in generated instructions.